src/v/cloud_storage/cache_service.cc

```diff
@@ -9,49 +9,160 @@
+        if (is_tmp_file(filename_to_remove)) {
+            bool successfully_deleted = true;
+            try {
+                co_await ss::remove_file(filename_to_remove);
+            } catch (std::exception& e) {
+                successfully_deleted = false;
+                vlog(
+                  cst_log.error,
+                  "Cache eviction couldn't delete {}: {}.",
+                  filename_to_remove,
+                  e.what());
+            }
+            if (successfully_deleted) {
+                _total_cleaned += file_item.size;
+            }
+        }
+        vlog(
+          cst_log.debug,
+          "Clean up at start deleted files of total size {}.",
+          _total_cleaned);
+    }
+}
+
+ss::future<> cache::clean_up_cache() {
+    gate_guard guard{_gate};
+    co_await _walker.walk(_cache_dir.native());
+
+    if (_walker.get_current_cache_size() >= _max_cache_size) {
+        auto size_to_delete = _walker.get_current_cache_size()
+                              - (_max_cache_size * _cache_size_low_watermark);
+
+        auto candidates_for_deletion = _walker.get_files();
+
+        uint64_t deleted_size = 0;
+        size_t i_to_delete = 0;
+        while (i_to_delete < candidates_for_deletion.size()
+               && deleted_size < size_to_delete) {
+            auto filename_to_remove
+              = candidates_for_deletion[i_to_delete].filename;
+            vassert(
+              0
+                == filename_to_remove.substr(0, _cache_dir.native().length())
+                     .compare(_cache_dir.native()),
+              "Tried to clean up {}, which is outside of cache_dir {}.",
+              filename_to_remove,
+              _cache_dir.native());
+
+            // skip tmp files since someone may be writing to it
+            if (!is_tmp_file(filename_to_remove)) {
+                bool successfully_deleted = true;
+                try {
+                    co_await ss::remove_file(filename_to_remove);
+                } catch (std::exception& e) {
+                    successfully_deleted = false;
+                    vlog(
+                      cst_log.error,
+                      "Cache eviction couldn't delete {}: {}.",
+                      filename_to_remove,
+                      e.what());
+                }
+                if (successfully_deleted) {
+                    deleted_size += candidates_for_deletion[i_to_delete].size;
+                }
+            }
+            i_to_delete++;
+        }
+        _total_cleaned += deleted_size;
+        vlog(
+          cst_log.debug,
+          "Cache eviction deleted {} files of total size {}.",
+          i_to_delete,
+          deleted_size);
+    }
+}
 
 ss::future<> cache::start() {
     vlog(cst_log.debug, "Starting archival cache service");
-    // todo: start maintenance cycle here
-    return ss::make_ready_future<>();
+    co_await clean_up_at_start();
```

Comment:
This should also be done at the shard 0, no?

Reply:
Fixed.